src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+import Data.Bits (Bits (..))
+import Data.ByteArray (ByteArrayAccess, convert)
+import Data.ByteString.Builder (Builder, byteString, lazyByteString)
+import Data.Foldable (Foldable (..))
+import Prelude (show)
+
+import qualified Data.ByteString.Builder.Extra as Builder
+import qualified Data.ByteString.Lazy as LBS
+
+-- | Data type for root of sized merkle tree.
+data SizedMerkleRoot a = SizedMerkleRoot
+    { smrHash :: Hash LBS.ByteString  -- ^ returns root 'Hash' of Merkle Tree
+    , smrSize :: Word8 -- ^ size of root node,
+                       -- size is defined as number of leafs in this subtree
+    } deriving (Show, Eq, Ord, Generic, ByteArrayAccess , Serialise, Typeable)
+
+
+data SizedMerkleTree a
+  = SizedMerkleEmpty
+  | SizedMerkleTree !(SizedMerkleNode a)
+  deriving (Eq, Generic)
+
+
+instance Foldable SizedMerkleTree where
+    foldMap _ SizedMerkleEmpty      = mempty
+    foldMap f (SizedMerkleTree n) = foldMap f n
+
+    null SizedMerkleEmpty = True
+    null _                = False
+
+    length SizedMerkleEmpty      = 0
+    length (SizedMerkleTree n) = fromInteger (toInteger (smrSize (mRoot n)))
+
+instance Show a => Show (SizedMerkleTree a) where
+  show SizedMerkleEmpty = "Empty Sized Merkle tree "
+  show (SizedMerkleTree n) = "Sized Merkle tree:  " <> Prelude.show n
+  --show tree = "Merkle tree: "  <> Prelude.show (toList tree)
+
+type LeafIndex = Int
+
+data SizedMerkleNode a
+    = SizedMerkleBranch { mRoot  :: !(SizedMerkleRoot a)
+                        , mLeft  :: !(SizedMerkleNode a)
+                        , mRight :: !(SizedMerkleNode a)}
+    | SizedMerkleLeaf { mRoot  :: !(SizedMerkleRoot a)
+                      , mIndex :: !LeafIndex
+                      , mVal   :: !a}
+    deriving (Eq, Show, Generic)
+
+instance Foldable SizedMerkleNode where
+  foldMap f x = case x of
+    SizedMerkleLeaf {mVal} -> f mVal
+    SizedMerkleBranch {mLeft, mRight} -> foldMap f mLeft `mappend` foldMap f mRight
+
+toLazyByteString :: Builder -> LBS.ByteString
+toLazyByteString = Builder.toLazyByteStringWith (Builder.safeStrategy 1024 4096) mempty
+
+mkLeaf :: HasHash a => (LeafIndex, a) -> SizedMerkleNode a
+mkLeaf (i, a) =
+    SizedMerkleLeaf
+    { mVal  = a
+    , mIndex = i
+    , mRoot = SizedMerkleRoot (unsafeHash a) -- unsafeHash since we need to hash to ByteString
+                              1 -- size of leaf node is 1
+    }
+
+mkBranch :: SizedMerkleNode a -> SizedMerkleNode a -> SizedMerkleNode a
+mkBranch l r =
+    SizedMerkleBranch
+    { mLeft  = l
+    , mRight = r
+    , mRoot  = mkBranchRootHash (mRoot l) (mRoot r)
+    }
+
+mkBranchRootHash :: SizedMerkleRoot a -- ^ left merkle root
+                 -> SizedMerkleRoot a -- ^ right merkle root
+                 -> SizedMerkleRoot a
+mkBranchRootHash (SizedMerkleRoot (AbstractHash hl) sl)
+                 (SizedMerkleRoot (AbstractHash hr) sr)
+  = SizedMerkleRoot
+     (hash $ toLazyByteString $ mconcat
+        [ byteString (one sl)
```

Comment:
Is it `byteString :: ByteString -> Builder`? Why not `mconcat $ map byteString $ sl : sr : (convert hl ++ convert hr)` then?

Also why do you even use a builder here?

Reply:
yes, it's `ByteString -> Builder`. This is how it is done in cardano-sl, I guess it's for performance reasons.

Reply:
I see, ok.